kernel/src/sched.rs

```diff
@@ -867,37 +889,99 @@ impl Kernel {
+                // process accordingly.
+                let fn_ptr: Option<NonNull<()>> = NonNull::new(upcall_ptr);
+
+                // A upcall is identified as a tuple of the driver
+                // number and the subdriver number.
                 let upcall_id = UpcallId {
                     driver_num: driver_number,
                     subscribe_num: subdriver_number,
                 };
-                // Only one upcall should exist per tuple.
-                // To ensure that there are no pending
-                // upcalls with the same identifier but
-                // with the old function pointer, we clear
-                // them now.
-                process.remove_pending_upcalls(upcall_id);
-
-                let ptr = NonNull::new(upcall_ptr);
-                let upcall = ptr.map_or(Upcall::default(), |ptr| {
-                    Upcall::new(process.processid(), upcall_id, appdata, ptr.cast())
-                });
-                let rval = platform.with_driver(driver_number, |driver| match driver {
+
+                // Construct the upcall struct
+                let upcall = Upcall::new(process.processid(), upcall_id, appdata, fn_ptr);
+
+                let rval = platform.with_driver(driver_number as usize, |driver| match driver {
                     Some(d) => {
-                        let res = d.subscribe(subdriver_number, upcall, process.processid());
+                        let res =
+                            d.subscribe(subdriver_number as usize, upcall, process.processid());
+
                         match res {
-                            // An Ok() returns the previous upcall, while
-                            // Err() returns the one that was just passed
-                            // (because the call was rejected).
-                            Ok(oldcb) => oldcb.into_subscribe_success(),
-                            Err((newcb, err)) => newcb.into_subscribe_failure(err),
+                            Err((returned_upcall, err)) => {
+                                // The capsule has refused the subscribe operation,
+                                // verify that it passed back the new upcall
+                                if returned_upcall.app_id != process.processid()
+                                    || returned_upcall.upcall_id != upcall_id
+                                    || returned_upcall.appdata != appdata
+                                    || returned_upcall.fn_ptr != fn_ptr
+                                {
+                                    // The capsule did not return the Upcall passed in
+                                    //
+                                    // TODO: How to handle this?
+                                    panic!(
+                                        "Driver {}, subscribe num {}: Upcall swapped in error case",
+                                        driver_number, subdriver_number
+                                    );
```

Comment:
This seems like an important todo. It's a shame if after all this work buggy capsules can still cause kernel panics.

Reply:
Replaced with terminating the process.